opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/drools/DroolsAlarmContext.java

```diff
@@ -89,68 +103,142 @@ public DroolsAlarmContext() {
+                    if (stateTracker.wasAlarmWithIdUpdated(alarmId)) {
+                        return false;
                     }
-                }
-            }
+                    final OnmsAlarm alarmInMem = alarmAndFact.getAlarm();
+                    final OnmsAlarm alarmInDb = alarmsInDbById.get(alarmId);
+                    // Only update the alarms if they are different
+                    return shouldUpdateAlarmForSnapshot(alarmInMem, alarmInDb);
+                })
+                .collect(Collectors.toSet());
 
-            for (Integer alarmIdToRemove : alarmIdsToRemove) {
-                handleDeletedAlarmNoLock(alarmIdToRemove);
-            }
-            for (Integer alarmIdToAdd : alarmIdsToAdd) {
-                handleNewOrUpdatedAlarmNoLock(alarmsInDbById.get(alarmIdToAdd));
+        // Log details that help explain what actions are being performed, if any
+        if (LOG.isDebugEnabled()) {
+            if (!alarmIdsToAdd.isEmpty() || !alarmIdsToRemove.isEmpty() || !alarmIdsToUpdate.isEmpty()) {
+                LOG.debug("Adding {} alarms, removing {} alarms and updating {} alarms for snapshot.",
+                        alarmIdsToAdd.size(), alarmIdsToRemove.size(), alarmIdsToUpdate.size());
+            } else {
+                LOG.debug("No actions to perform for alarm snapshot.");
             }
-            for (Integer alarmIdToUpdate : alarmIdsToUpdate) {
-                handleNewOrUpdatedAlarmNoLock(alarmsInDbById.get(alarmIdToUpdate));
+            // When TRACE is enabled, include diagnostic information to help explain why
+            // the alarms are being updated
+            if (LOG.isTraceEnabled()) {
+                for (Integer alarmIdToUpdate : alarmIdsToUpdate) {
+                    LOG.trace("Updating alarm with id={}. Alarm from DB: {} vs Alarm from memory: {}",
+                            alarmIdToUpdate,
+                            alarmsInDbById.get(alarmIdToUpdate),
+                            alarmsById.get(alarmIdToUpdate));
+                }
             }
+        }
 
-            LOG.debug("Done handling snapshot.");
-        } finally {
-            unlockIfNotFiring();
+        for (Integer alarmIdToRemove : alarmIdsToRemove) {
+            handleDeletedAlarmNoLock(alarmIdToRemove, alarmsById.get(alarmIdToRemove).getAlarm().getReductionKey());
+        }
+        for (Integer alarmIdToAdd : alarmIdsToAdd) {
+            handleNewOrUpdatedAlarmNoLock(alarmsInDbById.get(alarmIdToAdd));
+        }
+        for (Integer alarmIdToUpdate : alarmIdsToUpdate) {
+            handleNewOrUpdatedAlarmNoLock(alarmsInDbById.get(alarmIdToUpdate));
+        }
+
+        LOG.debug("Done handling snapshot.");
+    }
+
+    @Override
+    public void postHandleAlarmSnapshot() {
+        stateTracker.resetStateAndStopTrackingAlarms();
+        final ReentrantLock lock = getLock();
+        if (lock.isHeldByCurrentThread()) {
```

Comment:
Ok I see now. Looks like the try-finally could be rearranged to avoid this, ideally the call to post would never happen if the snapshot wasn't handled.